pxd.c

```diff
@@ -2153,9 +2159,8 @@ static void pxd_abort_context(struct work_struct *work)
 		ctx->name, ctx->id);
 
 	fc->allow_disconnected = 0;
-
 	/* Let other threads see the value of allow_disconnected. */
-	synchronize_rcu();
+	wmb();
```

Comment:
wmb by itself does not do much.

synchronize_rcu() has a stronger memory barrier guarantee. By the time synchronize_rcu() returns all threads including this one has done a memory barrier and will see the updated value of allow_disconnected.

Reply:
It did not happen. Earlier the usage was with synchronize_rcu() and read_[lock|unlock]_rcu. It did not ensure updated values get reflected in the IO enqueue path and that was the reason new requests got enqueued while abort handling was running. Also access to variables without READ/WRITE_ONCE macros allows those access to be reordered during execution.

In any case, with the posted changes the original issue of invalid index and assertion is not seen anymore.